Blog: Send a File to Node-RED (#1445)
FlowFuse/website: src/blog/2024/01/send-a-file.md

---
title: Send a File to Node-RED
subtitle: A guide to sending a csv file to Node-RED and start interacting with it.
date: 2024-01-03
authors: ["grey-dziuba"]
image: /blog/2023/12/images/uns-data-modeling.png
tags:
    - posts
    - flowfuse
    - how-to
---

> Have you ever needed to send a CSV file to your Node-RED instance? This file can go on to populate a shift schedule, product specifications, or some other configuration file that is used. In this guide, we are leveraging a Python script that runs a Flask application to post the file to a Node-RED instance and organize the data to be sent on or used.

Reviewer: @gdziuba Is `curl` available on windows? I think using curl is much easier for everyone.

Author: Yes, Windows 10 and above since 2018.  Manufacturing uses earlier versions of that though.  I can add it as a third option though.

Author: Yes.  Going to include.

<!--more-->

## Why would you need to send a file to Node-RED?

Often times it is necessary to update lookup tables in a SQL database, but you don't necessarily want to give access to everyone to edit the database, nor do you want to have to do it all yourself. This can often be seen when new products are introduced into a manufacturing facility. It may not be often, but enough that it warrants its own application. This process will guide you in a way that will enable your teammates to upload the files to the system themselves.

## Summary

> There are two approaches that were taken when solving this. The first approach is a simple script that will be shared below. It is a simple Python application that allows the user to send a file with a simple command, but this might require a little more technical skills that the end user may not feel comfortable with. Which brought us to pathway number two. A web-based application that allows the user to upload files to a browser with a selectable endpoint. Both have their pros and cons. I will let you decide which is best for your application.

Reviewer: Method one: `curl -X POST -d @/path/to/file.csv` seems very simple. For solution two; can we create a dashboard that allows for file uploads? (cc @joepavitt )

Reviewer 2: We have an issue for it as a core node: https://github.com/FlowFuse/node-red-dashboard/issues/249 although did also have someone that used `ui-template` to do it too: https://github.com/FlowFuse/node-red-dashboard/issues/369

Author: I am including both curl and a Dashboard input along the other two options.

# Simple Script

This Python script requires **requests** and **Python 3.x**.

Install requests:

```bash
pip install requests
```

### Configuration ###

Update the **nodered_url** to the location of the NodeRED instance.  Be sure to adjust the port if the default port of 1880 isn't being used.

Update the **file_path** with the path to where the file to be uploaded will be located.

> ```

Reviewer: ```suggestion
```python
```

---
title: Send a File to Node-RED
subtitle: A guide to sending a csv file to Node-RED and start interacting with it.

> date: 2024-01-04

Author: Why the 24th?

> date: 2024-01-04

Reviewer: The linked issue had a target date mentioning it, but tomorrow is fine.

> date: 2024-01-04

Author: Date format mismatch.

authors: ["grey-dziuba"]
image: /blog/2024/01/images/send-file-to-node-red.png
tags:
    - posts
    - flowfuse
    - how-to
---

Have you ever needed to send a CSV file to your Node-RED instance? This file can go on to populate a shift schedule, product specifications, or some other configuration file that is used. In this guide, we are leveraging a Python script that runs a Flask application to post the file to a Node-RED instance and organize the data to be sent on or used.
<!--more-->

# Why would you need to send a file to Node-RED?

Often times it is necessary to update lookup tables in a SQL database, but you don't necessarily want to give access to everyone to edit the database, nor do you want to have to do it all yourself. This can often be seen when new products are introduced into a manufacturing facility. It may not be often, but enough that it warrants its own application. This process will guide you in a way that will enable your teammates to upload the files to the system themselves.

# Summary

> There are many approaches that can be taken when solving this. We are going to go over 4. The first is the most simple approach but very manual.  A simple cURL command to send the file.  The second is to leverage FlowFuse's Dashboards to allow the user to navigate to the webpage upload a file and visualize before sending it on to the next process.  The third approach is a simple script that will be shared below. It is a simple Python application that allows the user to send a file with a simple command, but this might require a little more technical skills that the end user may not feel comfortable with. Which brought us to last option. A web-based application that allows the user to upload files to a browser with a selectable endpoint. All have their pros and cons. I will let you decide which is best for your application.

Reviewer: ```suggestion
There are many approaches that can be taken when solving this. We are going to go over 4. The first is the most simple approach but very manual. A simple cURL command to send the file. The second is to leverage [FlowFuse's Dashboard](https://dashboard.flowfuse.com/) to allow the user to navigate to the webpage upload a file and visualize before sending it on to the next process. The third approach is a simple script that will be shared below. It is a simple Python application that allows the user to send a file with a simple command, but this might require a little more technical skills that the end user may not feel comfortable with. Which brought us to last option. A web-based application that allows the user to upload files to a browser with a selectable endpoint. All have their pros and cons. I will let you decide which is best for your application.
```

## cURL

Honestly it doesn't get much more simple than this, but it is probably not for everyone.  Not everyone has access to a device that has cURL available.  Generally it is only available on Linux and UNIX.  It was announced to be available in Windows, but it isn't a direct translation and function more like a alias to "[Invoke-WebRequest](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-webrequest?view=powershell-7.4)".  It isn't a one to one.  None the less here is an example cURL command that will work with this [flow](#node-red-ingress).

```bash
curl -F 'data=@/path/to/file.csv' http://localhost:1880/fileupload
```

## NodeRED Dashboard (FlowFuse)

> This simple flow allows the user to visualize data from a CSV in FlowFuse Dashboards. The button then allows the user to initiate a request to send the data to the next step. This next step could be anything from loading into a SQL Database to saving it locally.

Reviewer: ```suggestion
This simple flow allows the user to visualize data from a CSV in FlowFuse Dashboard. The button then allows the user to initiate a request to send the data to the next step. This next step could be anything from loading into a SQL database to saving it.
```

![csv dashboard](./images/csv-dashboard.png)

<iframe width="100%" height="225px" src="https://flows.nodered.org/flow/8c505039ac1b8dbed2bee1e22ee2975a/share?height=100" allow="clipboard-read; clipboard-write" style="border: none;"></iframe>

> A link to the flow can be found [here](https://flows.nodered.org/flow/8c505039ac1b8dbed2bee1e22ee2975a).

Reviewer: If you click "Share" on the link you pasted there you can insert a flow and render it on the page.

```suggestion
<iframe width="100%" height="100%" src="https://flows.nodered.org/flow/8c505039ac1b8dbed2bee1e22ee2975a/share" allow="clipboard-read; clipboard-write" style="border: none;"></iframe>
```

To import the flow, follows these [instructions](https://flowfuse.com/blog/2023/03/3-quick-node-red-tips-5/#1.-copy-and-share-your-flows-using-export-and-import).

> ## Simple Script

Reviewer: Honestly I believe we should not include this; why complicate the article? IMO the dashboard method is best and is not explained well. The article should explain better why that works, and how it work. That's it for me, no python, no scripts. Even skip curl.

**Save**

To run:

```python
python run.py
```

## Web Application

![csv upload application](./images/csv_upload_app.png)

This stand alone web application can be run on either Windows or Linux, .bat for windows and .sh for linux.

### Installation

Clone the repository:
```bash
git clone https://github.com/gdziuba/FF_Send-File-to-NR.git
```
Navigate to the directory:
```bash
cd FF_Send-File-to-NR
```

### Configuration ###

Edit the lines in the body of [index.html](https://github.com/gdziuba/FF_Send-File-to-NR/blob/21214f88c6c4536f49efb88cf5f84bf52071a88b/templates/index.html#L69) to include the endpoints to which you would like to send the files.

```
<option value="http://localhost:1880/fileupload">CSV File Upload</option>
```

### Operating Systems

### Windows ###

Run the script:
```bash
.\start_app.bat
```

This will install if necessary, start the Flask Application, and take you to localhost:5000 on the browser.

### Linux ###

Make the script executable: Run:

```bash
chmod +x setup_and_run.sh
```

To Run:

```bash
./setup_and_run.sh
```

To access the application, open a browser to the ip:5000 of the running application.

> 
> 

Reviewer: ```suggestion

```

---

Have you ever needed to send a CSV file to your Node-RED instance? This file can go on to populate a shift schedule, product specifications, or some other configuration file that is used. In this guide, we provide multiple options to upload the data to your Node-RED for further processing and to organize the data to be sent on or used.

<!--more-->

# Why would you need to send a file to Node-RED?
Often times it is necessary to update lookup tables in a SQL database, but you don't necessarily want to give access to everyone to edit the database, nor do you want to have to do it all yourself. This can often be seen when new products are introduced into a manufacturing facility. It may not be often, but enough that it warrants its own application. This process will guide you in a way that will enable your teammates to upload the files to the system themselves.

Furthermore, on the management layer of most companies, Excel and Google Sheets are the go-to tools to perform data collection tasks. Getting management to involved in processes might require you to build an import feature for them. Asking your manager to "Save as" CSV is much easier than teaching them SQL!

> # Summary

Reviewer 3: ```suggestion
# 3 Ways to send a file to node-RED
```

There are many approaches that can be taken when solving this. We are going to go over three. 

1. [Node-RED Dashboard (FlowFuse)](#node-red-dashboard-(flowfuse)) - Leverage [FlowFuse's Dashboard](https://dashboard.flowfuse.com/) to allow the user to navigate to the webpage, upload a file, and visualize before sending it on to the next process. 
2. [Simple Python Script](#simple-python-script) - Simple script that will be shared below. It is a simple Python application that allows the user to send a file with a simple command, but this might require a little more technical skills that the end user may not feel comfortable with. 
3. [Stand Alone Web Application](#stand-alone-web-application) - A web-based application that allows the user to upload files to a browser with a selectable endpoint. All have their pros and cons. I will let you decide which is best for your application.

## Node-RED Dashboard (FlowFuse)

![csv dashboard](./images/csv-dashboard.png)

> This is the **recommended** approach.  This simple flow allows the user to visualize data from a CSV in the Node-RED Dashboard. The button then allows the user to initiate a request to send the data to the next step. This next step could be anything from loading into a SQL database to saving it.

Reviewer: ```suggestion
This is the **recommended** approach. This simple flow allows the user to visualize data from a CSV in the Node-RED Dashboard. The button then allows the user to initiate a request to send the data to the next step. This next step could be anything from loading into a SQL database to saving it.
```

### Instructions ###

> 1. Install Node-RED Dashboard 2.0.  Follow these [instructions](https://dashboard.flowfuse.com/getting-started.html) to install.  

Reviewer: ```suggestion
1. Install Node-RED Dashboard 2.0. Follow these [instructions](https://dashboard.flowfuse.com/getting-started.html) to install.  
```

2. Import Flow - to import the flow into your Node-RED instance follow these [instructions](https://flowfuse.com/blog/2023/03/3-quick-node-red-tips-5/#1.-copy-and-share-your-flows-using-export-and-import). 

> 3. Access Dashboard - To access the dashboard, navigate to the ```http://hostname:1880/dashboard```.

Reviewer: Can we assume people use FF Cloud? And just 
```suggestion
3. Access Dashboard - To access the dashboard, navigate to the `https://<flowfuse-instance-name>.flowfuse.cloud/dashboard`.
```

```python
import requests

def send_file(nodered_url, file_path):
    # Open the file in binary mode
    with open(file_path, 'rb') as file:
        files = {'file': (file.name, file, 'multipart/form-data')}
        response = requests.post(nodered_url, files=files)

    return response

# Update the ip address and port of your Node-RED instance
nodered_url = 'http://localhost:1880/fileupload'

# Update the location of your file
file_path = 'C:/Users/myUser/Downloads/shiftSchedule.csv'

response = send_file(nodered_url, file_path)
print(f"Response Status Code: {response.status_code}")
print(f"Response Body: {response.text}")
```

Update the **nodered_url** to the location of the NodeRED instance.  Be sure to adjust the port if the default port of 1880 isn't being used.

Update the **file_path** with the path to where the file to be uploaded will be located.

**Save**

To run:

```python
python run.py
```

## Stand Alone Web Application

![csv upload application](./images/csv_upload_app.png)

This stand-alone web application can be run on either Windows or Linux, .bat for Windows, and .sh for Linux.

### Installation

> Clone the repository:
> ```bash
> git clone https://github.com/gdziuba/FF_Send-File-to-NR.git
> ```
> Navigate to the directory:
> ```bash
> cd FF_Send-File-to-NR
> ```

Reviewer: ```suggestion
Clone the repository and navigate to the directory:
```bash
git clone https://github.com/gdziuba/FF_Send-File-to-NR.git && cd FF_Send-File-to-NR
```
```

### Configuration ###

Edit the lines in the body of [index.html](https://github.com/gdziuba/FF_Send-File-to-NR/blob/21214f88c6c4536f49efb88cf5f84bf52071a88b/templates/index.html#L69) to include the endpoints to which you would like to send the files.

```
<option value="http://localhost:1880/fileupload">CSV File Upload</option>
```

### Operating Systems

### Windows ###

Run the script:
```bash
.\start_app.bat
```

This will install if necessary, start the Flask Application, and take you to localhost:5000 on the browser.

### Linux ###

> Make the script executable: Run:

Reviewer: ```suggestion
Make the script executable by running running `chmod +x setup_and_run.sh`. Then run the application with: `chmod +x setup_and_run.sh`.
```

Reviewer: maybe this is more compact and better?

```bash
chmod +x setup_and_run.sh
```

To Run:

```bash
./setup_and_run.sh
```

> To access the application, open a browser to the ip:5000 of the running application.

Reviewer: ```suggestion
To access the application, open a browser to the <node-red-host-ip>:5000 of the running application.
```

### Node-RED Ingress

> <!-- ![csv upload application](./images/nr_flow_csv_ingress.png) -->

Reviewer: If we don't render it, remove this line and the image from this branch. If you don't know how to, I can too.

<iframe width="100%" height="225px" src="https://flows.nodered.org/flow/effb53752e5d6f767b3c7e5d41a4a6e8/share?height=100" allow="clipboard-read; clipboard-write" style="border: none;"></iframe>

> Once we have a file ready to be sent, we now need to configure the receiving side in Node-RED. In this example, we are leveraging a CSV formatted file and then converting it to be used at a later time.
> 
> A link to the flow can be found [here](https://flows.nodered.org/flow/effb53752e5d6f767b3c7e5d41a4a6e8).
> 
> To import the flow, follow these [instructions](https://flowfuse.com/blog/2023/03/3-quick-node-red-tips-5/#1.-copy-and-share-your-flows-using-export-and-import).
> 
> A Simple HTTP In node can be used in the form of a Post, ensuring the configuration allows for a file.

Reviewer: ```suggestion
Once we have a file ready to be sent, we now need to configure the receiving side in Node-RED. In this example, we are leveraging a CSV formatted file and then converting it to be used at a later time. A Simple HTTP In node can be used in the form of a Post, ensuring the configuration allows for a file.

To import the flow, follow these [instructions](https://flowfuse.com/blog/2023/03/3-quick-node-red-tips-5/#1.-copy-and-share-your-flows-using-export-and-import).
```

Reviewer: the iframe already has a link to copy. :)

Author: It does, but I am unsure if everyone knows how to use that.  That is why I have both.
